### ISSUE TITLE
"Variables '?firstName' are used in the projection result, but are not assigned."

### ISSUE THREAD

**Issue by RubenVerborgh** [comunica/comunica#670]
#### Issue type:

- :bug: Bug

____
#### Description:

```sparql
SELECT ?firstName WHERE {
  <https://rubenverborgh.inrupt.net/profile/card#me> <http://xmlns.com/foaf/0.1/knows> ?v0.
  ?v0 <http://xmlns.com/foaf/0.1/givenName> ?firstName.
}
```

on https://rubenverborgh.inrupt.net/profile/card recently started failing.

____
#### Environment:

| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.13.0
| node                | v12.14.0
| npm                 | 6.13.4
| yarn                | 1.22.0
| Operating System    | darwin (Darwin 18.7.0)

#### Crash log:

```
Error: Variables '?firstName' are used in the projection result, but are not assigned.
    at ActorQueryOperationProject.runOperation (/Users/ruben/Documents/UGent/Solid/solid-query-ldflex/node_modules/@comunica/actor-query-operation-project/lib/ActorQueryOperationProject.js:26:19)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async ActorQueryOperationProject.run (/Users/ruben/Documents/UGent/Solid/solid-query-ldflex/node_modules/@comunica/bus-query-operation/lib/ActorQueryOperationTyped.js:28:24)
    at async ActorInitSparql.query (/Users/ruben/Documents/UGent/Solid/solid-query-ldflex/node_modules/@comunica/actor-init-sparql/lib/ActorInitSparql-browser.js:90:24)
    at async ActorInitSparql.run (/Users/ruben/Documents/UGent/Solid/solid-query-ldflex/node_modules/@comunica/actor-init-sparql/lib/ActorInitSparql.js:141:29)
```

**Comment by joachimvh** [comment]
The problem can be found by looking [here](https://github.com/comunica/comunica/blob/master/packages/actor-query-operation-bgp-left-deep-smallest/lib/ActorQueryOperationBgpLeftDeepSmallest.ts#L206) and [here](https://github.com/comunica/comunica/blob/master/packages/actor-query-operation-project/lib/ActorQueryOperationProject.ts#L34).

More specifically: if a pattern has 0 results, the bgp actor will return an empty list of variables, which causes the project actor to complain since that makes them think the bindings do not contain results for the variables they're expecting. Solution would be to either make the bgp actor return the correct list of variables, or to make the project actor not check if there are no results anyway.

**Comment by joachimvh** [comment]
I should add that I didn't fully test the solutions here, but I did check and one of the patterns returned 0 results in its metadata so it seems quite plausible. (Not sure if the fact that one of the patterns has 0 results is another bug since I would expect Ruben's card to have results actually 😅 )

**Comment by rubensworks** [comment]
> make the bgp actor return the correct list of variables

This would probably be the cleanest solution.

**Comment by rubensworks** [comment]
Similar to #666

**Comment by RubenVerborgh** [comment]
> I would expect Ruben's card to have results actually

It's not my actual profile, but a test environment 🙂

**Comment by rubensworks** [comment]
Fixed in 1.13.1
